core/src/OpenACC/Kokkos_OpenACC.cpp

```diff
@@ -59,11 +70,52 @@ void Kokkos::Experimental::OpenACC::impl_initialize(
         acc_get_device_num(acc_device_host);
   } else {
     using Kokkos::Impl::get_visible_devices;
+    acc_set_device_type(Impl::OpenACC_Traits::dev_type);
     std::vector<int> const& visible_devices = get_visible_devices();
     using Kokkos::Impl::get_gpu;
     int const dev_num = get_gpu(settings).value_or(visible_devices[0]);
     acc_set_device_num(dev_num, Impl::OpenACC_Traits::dev_type);
     Impl::OpenACCInternal::m_acc_device_num = dev_num;
+#if defined(KOKKOS_IMPL_ARCH_NVIDIA_GPU)
+    struct cudaDeviceProp deviceProp;
+    cudaError error = cudaGetDeviceProperties(&deviceProp, dev_num);
+    if (error != cudaSuccess) {
+      std::ostringstream msg;
+      msg << "Error: During OpenACC backend initialization, failed to retrieve "
+          << "CUDA device properties: (" << cudaGetErrorName(error)
+          << "): " << cudaGetErrorString(error);
+      Kokkos::Impl::host_abort(msg.str().c_str());
+    }
+    Impl::OpenACCInternal::m_concurrency =
+        deviceProp.maxThreadsPerMultiProcessor * deviceProp.multiProcessorCount;
+#elif defined(KOKKOS_ARCH_AMD_GPU)
+    hipDeviceProp_t deviceProp;
+    hipError_t error = hipGetDeviceProperties(&deviceProp, dev_num);
+    if (error != hipSuccess) {
+      std::ostringstream msg;
+      msg << "Error: During OpenACC backend initialization, failed to retrieve "
+          << "HIP device properties: (" << hipGetErrorName(error)
+          << "): " << hipGetErrorString(error);
+      Kokkos::Impl::host_abort(msg.str().c_str());
+    }
+    Impl::OpenACCInternal::m_concurrency =
+        deviceProp.maxThreadsPerMultiProcessor * deviceProp.multiProcessorCount;
+#elif defined(KOKKOS_ENABLE_OPENACC_FORCE_HOST_AS_DEVICE)
+    Impl::OpenACCInternal::m_concurrency = std::thread::hardware_concurrency();
+    if (!Impl::OpenACCInternal::m_concurrency) {
+      Kokkos::Impl::host_abort(
+          "Error: During OpenACC backend initialization, failed to retrieve "
+          "CPU "
+          "hardware concurrency");
+    }
+#else
+    // FIXME_OPENACC: Compute Impl::OpenACCInternal::m_concurrency correctly.
+    if (show_warnings()) {
+      std::cerr << "Warning: OpenACC backend concurrency detection not "
+                   "implemented for target architecture"
+                << std::endl;
+    }
```

Comment:
Please confirm that you want to start raising warning in that case

Reply:
This part is for an unknown accelerator type (`acc_device_not_host`) where we don't know the concurrency.

Reply:
I understand that this is for the case where we use that 256000 fallback value.
I am just wondering if that is excessive and we would now be "spamming" the user about something that is not unexpectedly happening at runtime but that we can "diagnose" at configuration time and only warn then.
I don't feel too strongly about it if you want it though. Just observing that this somewhat departs from the usual situations in which we currently decide to warn users in the rest of the codebase.

Reply:
Removed the warning message; instead added a message in the CMake configuration (Kokkos_arch.cmake) regarding the automatic fallback mode.